mkcloud: support IPMI with libvirt nodes

cmurphy opened the pull request:
This adds a BMC interface to the libvirt node definition through QEMU, which allows IPMI kernel drivers to be loaded, thereby letting crowbar discover information about the BMC interface while the node is being joined. The BMC properties (such as its IP address and credentials) are then managed by binding an ipmi_sim instance to a veth pair that connects the BMC interface to the admin network bridge. This will allow us to properly test the IPMI barclamp in CI, and will also be useful for testing the ironic barclamp with production-grade drivers such as pxe_ipmitool.

bmwiedemann commented:
merged 1st commit. 2nd commit message has typo: s/tern/turn/

bmwiedemann commented on scripts/qa_crowbarsetup.sh:
```diff
@@ -1962,7 +1962,7 @@ function set_crowbar_ipmi_ip_addrs
     local bmc_net="$ip1.$ip2.$ip3"
     local i
-    for i in $(seq 1 $nodenumbertotal); do
+    for i in $(seq 1 $nodenumber); do
```
The nodenumbertotal here was meant to allow physical clouds of 5 nodes, where we only use 2 and turn off the other 3 via IPMI so that IP addrs assigned in the previous run do not collide with those of the active nodes.

bmwiedemann commented:
rebased it, added one commit to have a default nodenumbertotal value and added a fixup commit for using it again

cmurphy commented:
Squashed the fixup commit, fixed the hard-coded BMC credentials, and made it able to deploy the pacemaker barclamp with ipmi as the stonith device. Still need to work out the networking bits.

cmurphy commented:
Now uses the user-defined network instead of the default one for BMC addresses. The interface network config is read-only so the `use_dhcp` and `ignore_address_suggestions` barclamp params have no meaning, that's the next thing to fix.

cmurphy commented:
Added a commit to implement write functionality in the BMC simulator so we can have crowbar actually change the IP address of a BMC. Last thing to do is somehow make the use_dhcp option useful, but that code path in crowbar is actually pretty small so it's not as useful to implement as you might think.

dirkmueller commented on scripts/lib/mkcloud-ipmi.sh:
```diff
@@ -0,0 +1,45 @@
+    local password=$3
+    local listenaddr=$4
+    local nodename=$cloud-node$nodenumber
+    local lanconfig=/tmp/ipmi-lan-$nodename.conf
```
hardcoding in /tmp isn't very nice. is there a better place to put it? /var/run/mkcloud for example?

nicolasbock replied:
You could `lanlconfig=$(mktemp XXXX-impi-lan-$nodename.conf)`.

cmurphy replied:
@dirkmueller we already hardcode /tmp for libvirt vm definition files (e.g. https://github.com/SUSE-Cloud/automation/blob/b8dca9ad65a4b175cbf345d0995a85c6c82141b8/scripts/lib/mkcloud-libvirt.sh#L409) which are sort of analogous to this, so I thought it was appropriate here. I could use /var/run/mkcloud. It looks like the only thing using that directory is dirmaint, which I have no familiarity with, and I didn't want to trailblaze here. But if that's a better path forward then I can go with that.

@nicolasbock I didn't see the value of using mktemp here, I think it's better for the file name to be consistent. mktemp also defaults to putting files in /tmp so we'd still have to decide a better place for it.

nicolasbock replied:
@cmurphy I agree with you that it seems a bit like overkill to use `mktemp` here. In combination with exporting `TMPDIR` we could address @dirkmueller's concern regarding the location though. But then again, I also agree with you that we consistently use `/tmp` throughout, and it doesn't strike me as particularly unnice to use `/tmp`... @dirkmueller, your turn :smile:....

cmurphy replied:
I would be more concerned about using /tmp in combination with mktemp because of the possibility that things don't get cleaned up and /tmp fills up. By using a consistent name we can only have as many temp files as nodes that are running, so it won't grow exponentially.

nicolasbock replied:
Yes, that's a valid point. Something like `tempfile.TemporaryFile()` would be so nice now :smile:.

toabctl commented:
@cmurphy needs a rebase.

cmurphy commented:
@toabctl done

cmurphy commented:
Investigating whether I can replace the ipmi_sim parts with virtualbmc since it would be nice if this could be used to deploy nodes with ironic and I'm running into some incompatibilities with ironic and ipmi_sim (like https://sourceforge.net/p/openipmi/discussion/112995/thread/9135111d/, the fix for which isn't released in Leap yet)

cmurphy commented:
Virtualbmc really doesn't implement enough of the IPMI spec to support crowbar's needs, so it's not worth pursuing here (I'm remembering why I discounted it early on).

rsalevsky commented:
Needs a rebase

cmurphy commented:
@rsalevsky done

cmurphy commented:
@dirkmueller fixed

bmwiedemann commented on scripts/mkcloud:
```diff
@@ -326,7 +329,8 @@ function sshrun
         export ipmi_ip_addrs=$ipmi_ip_addrs;
-
+        export bmc_user=$bmc_user;
+        export bmc_password=$bmc_password;
```
should be moved to mkcloud-common, so that is also defined for those cases that run without mkcloud.

bmwiedemann commented on scripts/mkcloud:
```diff
@@ -173,6 +173,9 @@ host_mtu=$(determine_mtu)
     [[ -n $want_efi ]] && firmware_type="uefi"
 
+: ${bmc_user:=root}
+: ${bmc_password:=cr0wBar!}
```
should be moved into mkcloud-common so that it is also defined in mkphyscloud-like cases that run without mkcloud

cmurphy replied:
@bmwiedemann done

jdsn commented:
@dirkmueller please rereview or can we dismiss your review?
